### PR TITLE
Node Label Checkpoint Support

### PR DESCRIPTION
**Pull request by pwittrock**
Node Label Checkpoint Support
- Node labels specified only via directory containing static json / yaml files
  - Node labels no longer specified by single file or commandline list
- Changes to label files will be applied when the kubelet is restarted instead of only at creation time
  - Labels written to checkpoint file so they can be deleted when they are removed from the label files
  - Labels written on node status update so that they are written when the node is restarted
- Does not support updating labels in response to file changes without kubelet restart

### CODE REVIEW

**Comment by pwittrock** [comment]
Hey Vish,
Are you the best person to review this or should I assign to someone else?

**Comment by vishh** [comment]
What is the checkpoint format? How do we provide forwards/backwards compatibility?
Can we consider using protos?

**Comment by pwittrock** [comment]
The checkpoint format is a json file that matches the input file format.  We could use protos, or golang's native binary serialization if needed.  My main concern with doing so is that it makes things harder to debug since you can't open the checkpoint file and make sense of its contents.  Is there a precedent for using binary serialization elsewhere?  Are you thinking of reusing this file for checkpointing non-label data?

**Comment by vishh** [comment]
> The checkpoint format is a json file that matches the input file format. 

What is the input file format? Is it versioned?


> We could use protos, or golang's native binary serialization if needed. My main concern with doing so is that it makes things harder to debug since you can't open the checkpoint file and make sense of its contents. 

I agree. It's just that versioning with json is a bit complex. I'd be ok with checkpointing human friendly strings as long as its versioned and is future proof.

> Is there a precedent for using binary serialization elsewhere? Are you thinking of reusing this file for checkpointing non-label data?

I don't think binary is a necessity unless we are worried about the checkpointing becoming APIs :grin: 

We currently checkpoint quite a bit using docker labels. Arguably once this PR lands, we can move those labels to being in checkpoints.

Should we instead start with a generic kubelet checkpoint and have labels as the first entity in that checkpoint?

File: pkg/kubelet/labels.go [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+	if err != nil {
+		return err
+	}
+	return ioutil.WriteFile(path, b, 0644)
+}
+
+// readLabelsFile reads in and parses the yaml or json node labels file
+func (ls *labelSet) readLabelsFile(path string) (map[string]string, error) {
+	labels := make(map[string]string, 0)
+	kps := make(map[string]interface{}, 0)
```
Comment (vishh) [inline pkg/kubelet/labels.go]: This format needs to be formalized. How about using `KubeletConfiguration` Object here?

**Comment by vishh** [comment]
> Does not support updating labels in response to file changes without kubelet restart

Maybe we can consider adding dynamic updates of labels in the future.

**Comment by vishh** [comment]
Closing this PR for now @pwittrock

**Pull request closed without merging**